### PR TITLE
`fungibles` methods do not consume `AssetId`

### PR DESCRIPTION
**Pull request by dianpopa**
`Fungibles` methods now take `AssetId` by reference instead of value where the value is not needed.

`pallet-assets` methods now also take `AssetId` by reference instead of value where the value is not needed.

This avoids a lot of cloning across the codebase.

There are more objects and traits that could do the same, but this PR already blew up in size as I was changing the most relevant ones. This "small" change is actually pretty big..

fixes https://github.com/paritytech/polkadot-sdk/issues/3968

### CODE REVIEW

File: substrate/frame/support/src/traits/tokens/fungibles/hold.rs [hunk 1]
```diff
@@ -39,6 +39,7 @@ use sp_runtime::{DispatchError, DispatchResult, Saturating, TokenError};
 
 use super::*;
 
+// TODO
```
Comment (ggwpez) [inline substrate/frame/support/src/traits/tokens/fungibles/hold.rs:42]: ?

File: cumulus/parachains/runtimes/test-utils/src/lib.rs [hunk 2]
```diff
@@ -534,9 +534,9 @@ pub fn assert_metadata<Fungibles, AccountId>(
 	Fungibles: frame_support::traits::fungibles::metadata::Inspect<AccountId>
 		+ frame_support::traits::fungibles::Inspect<AccountId>,
 {
-	assert_eq!(Fungibles::name(asset_id.clone().into()), Vec::from(expected_name),);
-	assert_eq!(Fungibles::symbol(asset_id.clone().into()), Vec::from(expected_symbol),);
-	assert_eq!(Fungibles::decimals(asset_id.into()), expected_decimals);
+	assert_eq!(Fungibles::name(&asset_id.clone().into()), Vec::from(expected_name),);
+	assert_eq!(Fungibles::symbol(&asset_id.clone().into()), Vec::from(expected_symbol),);
```
Comment (liamaharon) [inline cumulus/parachains/runtimes/test-utils/src/lib.rs:538]: Possible to remove the `.clones` in scenarios like this or?

File: substrate/frame/support/src/traits/tokens/fungible/union_of.rs [hunk 3]
```diff
@@ -115,80 +115,80 @@ impl<
 	type AssetId = AssetKind;
 	type Balance = Left::Balance;
 
-	fn total_issuance(asset: Self::AssetId) -> Self::Balance {
-		match Criterion::convert(asset) {
+	fn total_issuance(asset: &Self::AssetId) -> Self::Balance {
+		match Criterion::convert(asset.clone()) {
```
Comment (liamaharon) [inline substrate/frame/support/src/traits/tokens/fungible/union_of.rs:119]: Hmm I wonder if `Criterion::convert` should also take a ref?

**Review by liamaharon — COMMENTED** [review]
Thanks for helping with this!

**Comment by gilescope** [comment]
Context: This is good clean up. It's because `AssetId` used to be `Copy` but we needed it to not be (we wanted to use `MultiLocation` as an AssetId but certainly for xcm v2 `MultiLocation` wasn't `Copy`).

File: cumulus/parachains/runtimes/test-utils/src/lib.rs (hunk 2 above)
Comment (dianpopa) [reply cumulus/parachains/runtimes/test-utils/src/lib.rs:538]: All these tests helper functions provide `asset_id` as `impl Into<Fungibles::AssetId> + Clone`. I would need to change the API of these functions, then a lot of test callsites.
I was thinking it's not worth the effort and the noise since these are only tests.

File: substrate/frame/support/src/traits/tokens/fungible/union_of.rs (hunk 3 above)
Comment (dianpopa) [reply substrate/frame/support/src/traits/tokens/fungible/union_of.rs:119]: I initially tried, but got lost in the sea of errors. There are so many generic types and trait bounds used here, hard to follow.

I will give it another try.

**Comment by dianpopa** [comment]
PR has _a lot_ of changes, but the vast majority is test code because of the changed inner functions APIs. The "main" API changes are small I think.

File: substrate/frame/support/src/traits/tokens/fungible/union_of.rs (hunk 3 above)
Comment (dianpopa) [reply substrate/frame/support/src/traits/tokens/fungible/union_of.rs:119]: Can't make it work.. got as far as https://github.com/dianpopa/polkadot-sdk/commit/090c74efa31d71604124ed3e660aba74e288d944, maybe someone can help or do it in another PR.

**Comment by dianpopa** [comment]
@liamaharon @gilescope is there anything else required here?

File: cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-rococo/src/tests/reserve_transfer.rs [hunk 4]
```diff
@@ -723,7 +723,10 @@ fn reserve_transfer_native_asset_from_system_para_to_para() {
 	let sender_balance_before = test.sender.balance;
 	let receiver_assets_before = PenpalA::execute_with(|| {
 		type ForeignAssets = <PenpalA as PenpalAPallet>::ForeignAssets;
-		<ForeignAssets as Inspect<_>>::balance(system_para_native_asset_location.clone(), &receiver)
+		<ForeignAssets as Inspect<_>>::balance(
+			&system_para_native_asset_location.clone(),
```
Comment (liamaharon) [inline cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-rococo/src/tests/reserve_transfer.rs:727]: Does this need `.clone()`?

**Review by liamaharon — APPROVED** [review]
Looks good, maybe some redundant `.clones` still in there?

**Comment by dianpopa** [comment]
@liamaharon is this fix for #3968 still wanted?
I already put in considerable time, I will only continue if actually needed..
If yes, can you help with reviewers?

**Comment by franciscoaguirre** [comment]
I agree this is a good fix and a tedious one (the amount of compiler errors, number of files), we should push for it to get merged before more conflicts pile up. @liamaharon thoughts?

**Comment by bkchr** [comment]
Looks like the pr is stale.
